### PR TITLE
Address issue with `FilterTag` matching

### PR DESCRIPTION
**Pull request by romancow**
Addressing issues mentioned by @brollb on commit https://github.com/webgme/webgme-taxonomy/commit/125f6c5ecfc75e853887c403fcc5606b00102b30 .

@brollb does this look correct? Also, is this check no longer needed in `isMatch`?
```typescript
let matched = (itemTag.ID === this.id) && (itemTag.value == this.value);
```

### CODE REVIEW

File: src/routers/Search/dashboard/src/Utils.ts [hunk 1]
```diff
@@ -105,6 +105,10 @@ export function encodeQueryParams(dict: {[key: string]: string}) {
     .join('&');
 }
 
-export function isObject(thing) {
+export function isObject<T>(thing: T): thing is Exclude<T, undefined | string | number | boolean | Array<any> | Function> {
```
Comment (brollb) [inline src/routers/Search/dashboard/src/Utils.ts:108]: why isn't this `thing: any`?

File: src/routers/Search/dashboard/src/Utils.ts [hunk 2]
```diff
@@ -105,6 +105,10 @@ export function encodeQueryParams(dict: {[key: string]: string}) {
...
   return typeof thing === 'object' && !Array.isArray(thing);
 }
+
+export function isDefined<T>(thing: T): thing is Exclude<T, undefined> {
```
Comment (brollb) [inline src/routers/Search/dashboard/src/Utils.ts:112]: (Same question as above)

File: src/routers/Search/dashboard/src/tags/FilterTag.ts [hunk 3]
```diff
@@ -38,39 +39,19 @@ export default class FilterTag<T extends string = string> {
...
+    const value = ItemTag.valueForKey(itemTag, this.id);
+    return isDefined(value);
   }
 
   isMatch(itemTag: ItemTag) {
-    const value = getItemTagValue(itemTag, this.id);
-    let matched = this.value ? this.value === value : !!value;
-    return matched;
+    if (!isDefined(this.value)) {
+      return true;
```
Comment (brollb) [inline src/routers/Search/dashboard/src/tags/FilterTag.ts:48]: This should probably check if the label (referring to a tag with no properties/fields) exists and not assume true.

File: src/routers/Search/dashboard/src/tags/ItemTag.ts [hunk 4]
```diff
@@ -1,11 +1,23 @@
...
-  export function valueForId(itemTag: ItemTag, id: string) {
-    return (itemTag.ID === id) ? itemTag.value : itemTag[id];
+  export function valueForKey(itemTag: ItemTag, key: string) {
+    if (itemTag.hasOwnProperty(key)) {
+      return itemTag[key];
+    }
+    let value: any;
+    Object.values(itemTag)
+      .filter(isObject)
+      .some(obj => {
```
Comment (brollb) [inline src/routers/Search/dashboard/src/tags/ItemTag.ts:16]: Why are you using `some` here? It appears it is being used like `forEach` since the return value is being ignored. Am I missing something?

File: src/routers/Search/dashboard/src/Utils.ts (hunk 2 above)
Comment (romancow) [reply src/routers/Search/dashboard/src/Utils.ts:112]: So if you have a type, say `Date | undefined`, `isDefined` can act as a type guard, letting typescript know the type after it's called. An example:
```typescript
const maybeDate = getDate(); // say getDate returns `Date | undefined`

if (isDefined(maybeDate)) {
    // typescript knows it's a Date object here
   return  maybeDate.toISOString();
}
else {
    // typescript knows it still could be undefined, so it's a type error
   return maybeDate.toISOString();
}
```

If `thing` is `any` instead of a generic, then it won't know it has to be a `Date` if `isDefined` returns true.

[Here](https://www.typescriptlang.org/play?#code/G4QwTgBAtiCeBGBTAIiALogXBVGIB8IBXAOwBNEAzASxMTICgHLSBjNagexImoGdkVWvQAUaABa0A5thAlYASmwTpvPhACiAD1YAbIhQA8c2ABpi5IXTIA+CAG8GEZxDCI0RMDzGSSUiACEALwQJES6ugoEhD6qQfGh4ZEA3AwAvkwsJOxcPPyCNNYA4oh0YNSshgAqNrF+2FVKECp+apo6+kZV5qQUhfR2ji6u7p7eLf7BiRFR+DETEPEhYTOpGQzUlBAi+VaiMAgo6IgKUUMuAPQXzbAADoh8rOW3aBB8HBHNvgDW6tSvrE44TIECQOGOBAsfWEjGGByQuEQADo0JwAJIAZQA8hi0OU-CIFGsmJttrt+mQSmUKiJ4UcMKcHE5Ltc0HcHk9qC8IN8SJwAO5-NAAcnUIHBeE48AAVoh2BBxIg3MznHTESj0djcfipITiYhdHxEEzhlcbvdHs9XryBUK3h9dBBAcDQcbensyOY+JxeCKxebjUqwJwwCroHAEccNZicXjpHr0kA) is a TS Playground that shows the type errors.

File: src/routers/Search/dashboard/src/Utils.ts (hunk 1 above)
Comment (romancow) [reply src/routers/Search/dashboard/src/Utils.ts:108]: Similar reason to the `isDefined` example below (though a little messier). Just trying to filter out the possibility that `thing` could be a non-Object if `isObject` returns true.

File: src/routers/Search/dashboard/src/tags/ItemTag.ts (hunk 4 above)
Comment (romancow) [reply src/routers/Search/dashboard/src/tags/ItemTag.ts:16]: It's just a way to have it stop iterating once one is found since it stops when it returns `true`, instead of iterating them all with `forEach`.

Comment (brollb) [reply src/routers/Search/dashboard/src/tags/ItemTag.ts:16]: Ah, gotcha. I almost never use `some` (or `every` for that matter :) ) and am pretty sure I have written `!!someArray.find(condFn);` when I could've used it. haha... Anyway, it looks like this could be simplified by using `find` (rather than relying on a side effect - setting `any` - and then returning that). Not a huge deal though.

Comment (romancow) [reply src/routers/Search/dashboard/src/tags/ItemTag.ts:16]: The problem with using `find` here is it returns the item in the array, not the value its function returned. Inside the function I'm setting `value` to a calculated value (returned by a recursive `valueForKey` call). So `find` isn't going to return that `valueForKey` value, just the top most array item.  I could do something like a `map` with `valueForKey` first, then `find`, but then again we get to potential unnecessary calls after we've already "found" the value we're looking for. Perhaps this won't ever be big enough to really make a huge difference there, but it's just kind of my gut to avoid it, especially when using recursion.

I could alternatively use a `for` loop, but thought I'd stick with the functional approach. Doesn't really matter to me, though. I think I've read that `for` loops are usually a tad faster than `Array` functions.

File: src/routers/Search/dashboard/src/tags/FilterTag.ts [hunk 5]
```diff
@@ -38,39 +39,18 @@ export default class FilterTag<T extends string = string> {
...
   }
 
   isMatch(itemTag: ItemTag) {
-    const value = getItemTagValue(itemTag, this.id);
-    let matched = this.value ? this.value === value : !!value;
-    return matched;
+    const value = ItemTag.valueForKey(itemTag, this.id);
+    return !isDefined(this.value)
+      ? isDefined(value)
+      : (this.value === value);
```
Comment (romancow) [inline src/routers/Search/dashboard/src/tags/FilterTag.ts:50]: @brollb Like this?

File: src/routers/Search/dashboard/src/Utils.ts (hunk 2 above)
Comment (brollb) [reply src/routers/Search/dashboard/src/Utils.ts:112]: ah, cool. I didn't realize TS supported this. Thanks :)

File: src/routers/Search/dashboard/src/tags/ItemTag.ts (hunk 4 above)
Comment (brollb) [reply src/routers/Search/dashboard/src/tags/ItemTag.ts:16]: Makes sense!

File: src/routers/Search/dashboard/src/tags/FilterTag.ts (hunk 5 above)
Comment (brollb) [reply src/routers/Search/dashboard/src/tags/FilterTag.ts:50]: Yeah, I think that should be correct. Maybe this will better convey my thoughts:
```javascript
const isLabelTag = !isDefined(this.value);
const foundTag = isDefined(value);
const tagValuesMatch = (this.value === value);
return isLabelTag ? foundTag : tagValuesMatch;
```

Not saying you need to use this code - just trying to share my thoughts about the logical flow I was trying to capture.

**Review by brollb — APPROVED** [review]
I think this looks good. It would be nice to have a testing framework set up for this so we can verify these edge cases with tests. (It would be easier to verify than simply going off of my comments :) )

File: src/routers/Search/dashboard/src/tags/FilterTag.ts (hunk 5 above)
Comment (romancow) [reply src/routers/Search/dashboard/src/tags/FilterTag.ts:50]: Yeah, that is a bit more "self-documenting" with the variable names. We could take it a slight step further and add a `isLabel` property.
